communications/consumers.py

```
@@ -0,0 +1,160 @@
+        elif not self.user.is_authenticated:
+            error = f"{self.user} is not authenticated."
+
+        # Check both companies exist
+        elif not self.company1 or not self.company2:
+            error = "One or both companies are missing."
+
+        # Ensure exactly one company is a STARTUP
+        else:
+            is_startup_1 = self.company1.type == CompanyType.STARTUP
+            is_startup_2 = self.company2.type == CompanyType.STARTUP
+
+            if is_startup_1 and is_startup_2:
+                error = "Both companies are STARTUPs — only one is allowed."
+            elif not is_startup_1 and not is_startup_2:
+                error = "Neither company is a STARTUP — one must be."
+
+            # Swap so that company2 is always the STARTUP
+            elif is_startup_1 and not is_startup_2:
+                self.company1, self.company2 = self.company2, self.company1
+
+        return error
+
+    async def get_company_by_id(self, company_id):
+        try:
+            return await sync_to_async(CompanyProfile.objects.get)(id=company_id)
+        except CompanyProfile.DoesNotExist:
+            return None
+
+    async def get_or_create_chat_room(self, company1, company2):
+        """
+        Creates or retrieves a chat room between two users.
+        """
+        try:
+            room, _ = await sync_to_async(ChatRoom.objects.get_or_create)(
+                company_id_1=company1, company_id_2=company2
+            )
+            return room
+        except IntegrityError:
+            # In case of conflicts, retrieve the existing room
+            return await sync_to_async(ChatRoom.objects.get)(
+                company_id_1=company1, company_id_2=company2
+            )
+
+    async def disconnect(self, close_code):
+        """
+        Disconnects the user from the chat group upon exit.
+        """
+        await self.channel_layer.group_discard(self.room_group_name, self.channel_name)
```

If `connect()` fails early due to validation, `self.room_group_name` might not exist during disconnect.
```
    async def disconnect(self, close_code):
        if hasattr(self, "room_group_name"):
            await self.channel_layer.group_discard(self.room_group_name, self.channel_name)
```

---

Rly, a very good advice, thx